GUI/Types/Renderer/PhysSceneNode.cs

```diff
@@ -7,31 +9,72 @@ namespace GUI.Types.Renderer
 {
     class PhysSceneNode : ShapeSceneNode
     {
+        private static readonly Color32 ColorSphere = new(0f, 1f, 0f, 0.65f);
+        private static readonly Color32 ColorCapsule = new(0f, 1f, 0f, 0.65f);
+        private static readonly Color32 ColorMesh = new(0f, 0f, 1f, 0.65f);
+        private static readonly Color32 ColorHull = new(1.0f, 0.0f, 0.0f, 0.65f);
+
         public override bool LayerEnabled => Enabled && base.LayerEnabled;
         public bool Enabled { get; set; }
         public string PhysGroupName { get; set; }
 
-        public PhysSceneNode(Scene scene, List<SimpleVertex> verts, List<int> inds) : base(scene, verts, inds)
+        public PhysSceneNode(Scene scene, List<SimpleVertexNormal> verts, List<int> inds) : base(scene, verts, inds)
         {
         }
 
 
         public static IEnumerable<PhysSceneNode> CreatePhysSceneNodes(Scene scene, PhysAggregateData phys, string fileName)
         {
             var groupCount = phys.CollisionAttributes.Count;
-            var verts = new List<SimpleVertex>[groupCount];
+            var verts = new List<SimpleVertexNormal>[groupCount];
             var inds = new List<int>[groupCount];
             var boundingBoxes = new AABB[groupCount];
             var boundingBoxInitted = new bool[groupCount];
 
+
+            // compute space needed for all verts and indices
+            var indsSizes = new int[groupCount];
+            var vertsSizes = new int[groupCount];
+
+            var hemisphereVerts = SphereBands * SphereSegments + 1;
+            var hemisphereTriangles = SphereSegments * (2 * SphereBands - 1);
+            var capsuleTriangles = 2 * hemisphereTriangles + 2 * SphereSegments;
+
+            for (var p = 0; p < phys.Parts.Length; p++)
+            {
+                var shape = phys.Parts[p].Shape;
+
+                foreach (var sphere in shape.Spheres)
+                {
+                    vertsSizes[sphere.CollisionAttributeIndex] += hemisphereVerts;
+                    indsSizes[sphere.CollisionAttributeIndex] += hemisphereTriangles * 6;
+                }
+                foreach (var capsule in shape.Capsules)
+                {
+                    indsSizes[capsule.CollisionAttributeIndex] += capsuleTriangles * 6;
+                    vertsSizes[capsule.CollisionAttributeIndex] += hemisphereVerts * 2;
+                }
+                foreach (var hull in shape.Hulls)
+                {
+                    var numTriangles = hull.Shape.GetEdges().Length - hull.Shape.GetFaces().Length * 2;
```

Comment:
I verified that calculated sizes are correct, however calling GetEdges and GetTriangles is less than ideal unfortunately.